Feature: Implement a REPL

CompeyDev opened the pull request:
As requested by users before, this PR includes a REPL functionality which is invoked when Lune is run with no positional arguments. The REPL currently supports
- Evaluation
- Clean exits (^C & ^D)
- Execution history 
- ~~Preservation of code "context" (variables declared in previous steps)~~
   - This has proven to be tough and should have a separate PR entirely dedicated to it. There are simply too many moving parts and is therefore out of scope for now.

What needs to be implemented/fixed:
- [x] ~~Context preservation for other declaratives (such as functions, or globals)~~
- [x] ~~Global variable detections for pretty printing~~
- [x] Allow user to enter additional code if the statement is incomplete

Closes #38.

filiptibell commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,140 @@
+
+    let mut interrupt_counter = 0u32;
+
+    let colorize: Lazy<bool> = Lazy::new(|| {
+        let no_color = env::var("NO_COLOR").unwrap_or_else(|_| "false".to_string());
+
+        if no_color.is_empty() {
+            true
+        } else {
+            !env_var_bool(no_color).unwrap_or_else(|| false)
+        }
+    });
```
I think we can probably just default to always having colors on instead of this env var. The flag in `pretty_format_luau_error` is there to force colorization to be off in cases like stringifying errors from within lua (e.g. `tostring`), but for the CLI the `console` library we use for colorization will check for interactive terminals etc. and handle colors on/off there in a standardized way

CompeyDev replied:
Ah, that's great.

filiptibell commented on src/lib.rs:
```diff
@@ -6,4 +6,4 @@ pub mod roblox;
 mod tests;
 
-pub use crate::lune::{Lune, LuneError};
+pub use crate::lune::{lua, Lune, LuneError};
```
I don't think we should export the entire `lua` module, this would more or less export the entire Lune runtime and builtin libraries and if we change any of it we break semver compatibility for lib consumers due to it being exported.

```suggestion
pub use crate::lune::{Lune, LuneError};
```

CompeyDev replied:
How exactly do I go about using `lua::formatting::*` in repl.rs?

filiptibell replied:
My bad, I didn't notice that we were actually dealing with a `LuneError` and not a `LuaError`, you can actually remove the dependency on the formatting module & function completely here since `LuneError` is just an opaque and already formatted `LuaError`, and print it out directly:
```rs
eprintln!("{err}") // err is a LuneError and implements Display
```

filiptibell reviewed (changes requested):
Thank you for the PR! This looks good overall, I left a few comments to resolve.

RE: "Allow user to enter additional code if the statement is incomplete" I think following the example in the `mlua` repository would work well here too: https://github.com/khvzak/mlua/blob/0e4476c2e3be679552b110add270b2fd0f8fd2a0/examples/repl.rs#L33-L40

u-train commented:
An inquiry as a passer-by, would it be possible to have the REPL as a Lune application instead of it being implemented in Rust? It would make things easier by the looks of it, and it would be dogfooding Lune to itself.

CompeyDev commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,140 @@
+            dir.join(".lune_history")
+        } else {
+            eprintln!("Failed to find user home directory, abort!");
+            // Doesn't feel right to exit directly with a exit code of 1
+            // Lmk if there is a better way of doing this
```
I'm aware, it just won't traverse upwards the stack as its not within the scope of the direct show_repl function... right?

CompeyDev replied:
Nevermind, I'm stupid...

filiptibell replied:
You're right, I didn't see the closure here. How about something like:

```rs
let history_path = match UserDirs::new() {
    Some(dirs) => dirs.home_dir().join(".lune_history"),
    None => {
        eprintln!("Failed to find user home directory!");
        return Ok(ExitCode::FAILURE);
    }
}

match repl.load_history(&history_path, ...
```

Then you could also use this pathbuf for saving history below and probably replace usage of `save_repl_activity` with a plain `repl.save_history(&history_path)?`

CompeyDev commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,140 @@
+                }
+            }
+
+            eprintln!("WARN: Failed to load REPL history")
```
On the first run, we create the file - but we cannot load the file into our already created REPL instance. So instead we warn the user.

If the message seems misleading, I could update it to something like "WARN: First REPL session, could not load REPL history. Try restarting lune".

CompeyDev replied:
> How do other REPL environments deal with this? Do they just always create a history file on startup if one does not exist? Can we check if one exists before loading history and then create it, removing this error case check?

We could do that; totally didn't realize the simplest solution :joy:

CompeyDev commented:
> An inquiry as a passer-by, would it be possible to have the REPL as a Lune application instead of it being implemented in Rust? It would make things easier by the looks of it, and it would be dogfooding Lune to itself.

As per various standards, REPLs are the in-built into interpreters, for quick code execution.

CompeyDev commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,140 @@
+}
+
+// Isn't dependency injection plain awesome?!
+pub async fn show_interface(cmd: Command) -> Result<ExitCode, Error> {
```
👍

CompeyDev commented:
A question: As per the [mlua REPL example](https://github.com/khvzak/mlua/blob/0e4476c2e3be679552b110add270b2fd0f8fd2a0/examples/repl.rs#L34), they check whether the error is for an "incomplete input" as such. 

In the case of lune, a `LuneError` is returned, which does not provide details about the error at all. How do I go about getting the type of error it is?

filiptibell commented:
> In the case of lune, a `LuneError` is returned, which does not provide details about the error at all. How do I go about getting the type of error it is?

I added `is_incomplete_input` to `LuneError` in https://github.com/filiptibell/lune/commit/0ab32ffffd118e9dfa4fd51c4aaf3f41689ac914 , I think this should help

CompeyDev commented:
> `is_incomplete_input`

Awesome! Will implement.

CompeyDev commented:
I've implemented the input continuation feature, lmk if any changes required regarding that.

CompeyDev commented:
The only thing I see unsatisfactory at the moment is [this](https://github.com/filiptibell/lune/blob/aa50d3e5af7e4439cb397530c71accb0f12195a1/src/cli/repl.rs#L93) direct exit. I'll see what I can do to remove its usage.

filiptibell commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,114 @@
+        .join(".lune_history");
+
+    if !history_file_path.exists() {
+        std::fs::write(
+            // We know for sure that the home dir already exists
+            directories::UserDirs::new()
+                .unwrap()
+                .home_dir()
+                .join(".lune_history"),
+            String::new(),
+        )?;
```
```suggestion
        std::fs::write(&history_file_path, String::new())?;
```

CompeyDev replied:
I smooth-brained out here again 😂

filiptibell commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,114 @@
+                break;
+            }
+            Err(err) => {
+                eprintln!("REPL ERROR: {}", err.to_string());
+
+                // This isn't a good way to exit imo, once again
+                exit(1);
```
```suggestion
                eprintln!("REPL ERROR: {err}");
                return Ok(ExitCode::FAILURE);
```

CompeyDev replied:
Totally didn't just realize that I could just return the value returned by the match statement, my bad!

filiptibell commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,114 @@
+            Err(err) => {
+                if err.is_incomplete_input() {
+                    prompt_state = PromptState::Continuation;
+                    source_code.push_str("\n")
```
```suggestion
                    source_code.push('\n')
```

filiptibell commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,114 @@
+                    prompt_state = PromptState::Continuation;
+                    source_code.push_str("\n")
+                } else {
+                    eprintln!("{}", err);
```
```suggestion
                    eprintln!("{err}");
```

CompeyDev replied:
fix: inline error formatting (x2)

filiptibell commented on src/cli/repl.rs:
```diff
@@ -0,0 +1,114 @@
+        match repl.readline(prompt) {
+            Ok(code) => {
+                if prompt_state == PromptState::Continuation {
+                    write!(&mut source_code, "{}", code)?;
```
```suggestion
                    source_code.push_str(&code);
```

filiptibell commented:
> I've implemented the input continuation feature, lmk if any changes required regarding that.

Looks good to me!

> The only thing I see unsatisfactory at the moment is [this](https://github.com/filiptibell/lune/blob/aa50d3e5af7e4439cb397530c71accb0f12195a1/src/cli/repl.rs#L93) direct exit. I'll see what I can do to remove its usage.

I added some more review comments that you should be able to just click accept on, after this I think CI should also pass and we should be good to go

CompeyDev commented:
> > I've implemented the input continuation feature, lmk if any changes required regarding that.
> 
> Looks good to me!
> 
> > The only thing I see unsatisfactory at the moment is [this](https://github.com/filiptibell/lune/blob/aa50d3e5af7e4439cb397530c71accb0f12195a1/src/cli/repl.rs#L93) direct exit. I'll see what I can do to remove its usage.
> 
> I added some more review comments that you should be able to just click accept on, after this I think CI should also pass and we should be good to go

Gone ahead and accepted the changes, apologies for my stupidity at some times! 😅

fewkz commented:
Without "context", this isn't really that helpful. I fear the "adding context support later" will never end up being added and we'll be stuck with a pointless repl. You can **already** run lune code from the command line by doing `lune -`, so this doesn't exactly add anything new. At the very least, I'd prefer my PR to not be closed.

filiptibell commented:
> Without "context", this isn't really that helpful.

Context preservation would be blocked right now either way since there is no way for the REPL to get/set the function environment of what it wants to run. I'm in the process of rewriting the Lune scheduler right now and when that's done I think we will be able to integrate the REPL a bit more and get context preservation working. If you want to still keep the issue open though feel free to reopen it or even make a new issue for context preservation

CompeyDev commented:
> Without "context", this isn't really that helpful. I fear the "adding context support later" will never end up being added and we'll be stuck with a pointless repl. You can **already** run lune code from the command line by doing `lune -`, so this doesn't exactly add anything new. At the very least, I'd prefer my issue to not be closed.

As for now, you could use one-liners separated by semi-colors (even though that's not very convenient), and it will serve the same purposes. At the moment, the REPL is an MVP, similar to what is already available for Lua 5.4 REPL.
